File: backend/scripts/add_swagger_dto_props.js
const fs = require('fs');
const path = require('path');

const root = path.join(process.cwd(), 'src');
const files = [];

function findDtoFiles(dir) {
  const items = fs.readdirSync(dir);
  for (const item of items) {
    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);
    if (stat.isDirectory()) {
      findDtoFiles(fullPath);
    } else if (item.endsWith('.dto.ts')) {
      files.push(fullPath);
    }
  }
}

findDtoFiles(root);

const propPattern = /^(?<indent>\s*)(?<access>public |private |protected )?(?<readonly>readonly )?(?<name>[\w$]+)(?<optional>\?)?: (?<type>[^;]+);\s*$/;

for (const filePath of files) {
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split('\n');
  let modified = false;
  let apiImportLine = null;
  let lastImportIdx = -1;

  for (let idx = 0; idx < lines.length; idx++) {
    const line = lines[idx];
    if (line.startsWith('import ')) {
      lastImportIdx = idx;
    }
    if (line.includes("from '@nestjs/swagger'") && line.includes('ApiProperty')) {
      apiImportLine = idx;
    }
  }

  if (apiImportLine === null) {
    if (lastImportIdx >= 0) {
      lines.splice(lastImportIdx + 1, 0, "import { ApiProperty } from '@nestjs/swagger';");
      modified = true;
    } else {
      lines.unshift("import { ApiProperty } from '@nestjs/swagger';");
      modified = true;
    }
  }

  let idx = 0;
  let inConstructor = false;
  let methodBraceDepth = 0;

  while (idx < lines.length) {
    const line = lines[idx];
    const stripped = line.trim();

    if (!inConstructor && (stripped.includes('constructor(') || (stripped.match(/.+\(.*\)\s*{$/) && !stripped.startsWith('@')))) {
      inConstructor = true;
      methodBraceDepth = stripped.split('{').length - stripped.split('}').length;
      idx++;
      continue;
    } else if (inConstructor) {
      methodBraceDepth += stripped.split('{').length - stripped.split('}').length;
      if (methodBraceDepth <= 0) {
        inConstructor = false;
      }
      idx++;
      continue;
    }

    const match = propPattern.exec(line);
    if (match) {
      let start = idx;
      while (start - 1 >= 0 && lines[start - 1].trim().startsWith('@')) {
        start--;
      }

      const hasApi = lines.slice(start, idx).some(l => l.includes('ApiProperty'));
      if (!hasApi) {
        const optional = (match.groups.optional === '?' || lines.slice(Math.max(0, start - 3), idx).some(l => l.includes('@IsOptional')));
        const name = match.groups.name;
        const type = match.groups.type.trim();
        const desc = `${name} field`;

        let example = 'example';
        if (name.toLowerCase().includes('id') && !['idtoken', 'identity'].includes(name.toLowerCase())) {
          example = '123e4567-e89b-12d3-a456-426614174000';
        } else if (name.toLowerCase() === 'email') {
          example = 'user@example.com';
        } else if (name.toLowerCase().includes('password')) {
          example = 'P@ssw0rd123';
        } else if (name.toLowerCase().includes('url') || name.toLowerCase().includes('uri')) {
          example = 'https://example.com/resource';
        } else if (name.toLowerCase() === 'title') {
          example = 'Intro to Blockchain';
        } else if (name.toLowerCase() === 'description') {
          example = 'A concise description of the resource.';
        } else if (['firstname', 'lastname', 'username', 'name'].includes(name.toLowerCase())) {
          example = 'Jane Doe';
        } else if (['amount', 'score', 'xp', 'balance', 'limit', 'page', 'quantity'].includes(name.toLowerCase()) ||
                   type.toLowerCase().includes('number') || type.toLowerCase().includes('int') || type.toLowerCase().includes('float')) {
          example = 1;
        } else if (type.toLowerCase().includes('boolean') || name.toLowerCase().startsWith('is') || name.toLowerCase().startsWith('has')) {
          example = true;
        } else if (type.toLowerCase().includes('date')) {
          example = '2026-04-22T00:00:00.000Z';
        } else if (type.includes('[]') || type.toLowerCase().includes('array')) {
          example = ['example'];
        }

        const exampleRepr = typeof example === 'string' ? `'${example}'` : example;
        const requiredPart = optional ? ', required: false' : '';
        const apiDecorator = `${match.groups.indent}@ApiProperty({ example: ${exampleRepr}, description: '${desc}'${requiredPart} })`;

        lines.splice(start, 0, apiDecorator);
        idx++;
        modified = true;
      }
    }

    idx++;
  }

  if (modified) {
    fs.writeFileSync(filePath, lines.join('\n') + '\n');
    console.log(`Updated ${filePath}`);
  }
}

File: backend/scripts/add_swagger_dto_props.py
import re
from pathlib import Path

root = Path('backend/src')
files = sorted(root.rglob('*.dto.ts'))
prop_pattern = re.compile(r'^(?P<indent>\s*)(?P<access>public |private |protected )?(?P<readonly>readonly )?(?P<name>[\w$]+)(?P<optional>\?)?: (?P<type>[^;]+);\s*$')

for path in files:
    lines = path.read_text().splitlines()
    modified = False
    api_import_line = None
    last_import_idx = -1

    for idx, line in enumerate(lines):
        if line.startswith('import '):
            last_import_idx = idx
        if "from '@nestjs/swagger'" in line and 'ApiProperty' in line:
            api_import_line = idx

    if api_import_line is None:
        if last_import_idx >= 0:
            lines.insert(last_import_idx + 1, "import { ApiProperty } from '@nestjs/swagger';")
            modified = True
        else:
            lines.insert(0, "import { ApiProperty } from '@nestjs/swagger';")
            modified = True

    idx = 0
    in_constructor = False
    method_brace_depth = 0

    while idx < len(lines):
        line = lines[idx]
        stripped = line.strip()

        if not in_constructor and ('constructor(' in stripped or (re.match(r'.+\(.*\)\s*{$', stripped) and not stripped.startswith('@'))):
            in_constructor = True
            method_brace_depth = stripped.count('{') - stripped.count('}')
            idx += 1
            continue
        elif in_constructor:
            method_brace_depth += stripped.count('{') - stripped.count('}')
            if method_brace_depth <= 0:
                in_constructor = False
            idx += 1
            continue

        m = prop_pattern.match(line)
        if m:
            start = idx
            while start - 1 >= 0 and lines[start-1].lstrip().startswith('@'):
                start -= 1

            has_api = any('ApiProperty' in lines[i] for i in range(start, idx))
            if not has_api:
                optional = ('?' in m.group('optional') or any('@IsOptional' in lines[i] for i in range(max(0, start-3), idx)))
                name = m.group('name')
                t = m.group('type').strip()
                desc = f'{name} field'

                if 'id' in name.lower() and name.lower() not in ('idtoken', 'identity'):
                    example = '123e4567-e89b-12d3-a456-426614174000'
                elif name.lower() == 'email':
                    example = 'user@example.com'
                elif 'password' in name.lower():
                    example = 'P@ssw0rd123'
                elif 'url' in name.lower() or 'uri' in name.lower():
                    example = 'https://example.com/resource'
                elif name.lower() == 'title':
                    example = 'Intro to Blockchain'
                elif name.lower() == 'description':
                    example = 'A concise description of the resource.'
                elif name.lower() in ('firstname', 'lastname', 'username', 'name'):
                    example = 'Jane Doe'
                elif name.lower() in ('amount', 'score', 'xp', 'balance', 'limit', 'page', 'quantity') or 'number' in t.lower() or 'int' in t.lower() or 'float' in t.lower():
                    example = 1
                elif 'boolean' in t.lower() or name.lower().startswith('is') or name.lower().startswith('has'):
                    example = True
                elif 'date' in t.lower():
                    example = '2026-04-22T00:00:00.000Z'
                elif '[]' in t or 'array' in t.lower():
                    example = ['example']
                else:
                    example = 'example'

                if example is True:
                    example_repr = 'true'
                elif example is False:
                    example_repr = 'false'
                else:
                    example_repr = repr(example)

                required_part = '' if not optional else ', required: false'
                api_decorator = f"{m.group('indent')}@ApiProperty({{ example: {example_repr}, description: '{desc}'{required_part} }})"
                lines.insert(start, api_decorator)
                idx += 1
                modified = True

        idx += 1

    if modified:
        path.write_text('\n'.join(lines) + '\n')
        print(f'Updated {path}')
